Data.Replicator.Linear.Internal doesn't compile under GHC 9.2

**Describe the bug**

```
> cabal configure --with-ghc=ghc-9.2
...
> cabal build
...
src/Data/Replicator/Linear/Internal.hs:169:18: error:
    • Could not deduce (Elim (Arity b (a %1 -> f)) a b f1)
        arising from a use of ‘elim'’
      from the context: (n ~ Arity b (a %1 -> f1), Elim (n - 1) a b f1)
        bound by the instance declaration
        at src/Data/Replicator/Linear/Internal.hs:166:31-101
      The type variable ‘f’ is ambiguous
      Relevant bindings include
        r' :: Replicator a
          (bound at src/Data/Replicator/Linear/Internal.hs:169:11)
        a :: a (bound at src/Data/Replicator/Linear/Internal.hs:169:8)
        r :: Replicator a
          (bound at src/Data/Replicator/Linear/Internal.hs:167:11)
        g :: a %1 -> f1
          (bound at src/Data/Replicator/Linear/Internal.hs:167:9)
        elim' :: (a %1 -> f1) %1 -> Replicator a %1 -> b
          (bound at src/Data/Replicator/Linear/Internal.hs:167:3)
    • In the expression: elim' (g a) r'
      In a case alternative: (a, r') -> elim' (g a) r'
      In the second argument of ‘(&)’, namely
        ‘\case (a, r') -> elim' (g a) r'’
    |
169 |       (a, r') -> elim' (g a) r'
    |                  ^^^^^
...
```

**Environment**
 - OS name + version:
 - Version of the code: master branch

**Additional context**
Add any other context about the problem here.

--- a/shell.nix
+++ b/shell.nix
@@ -19,6 +19,9 @@ let
         "
     '';
   };
+  brokenIn92 = if ghcVersion == "921" then [] else [
+    pkgs.haskell-language-server
+  ];
 in with pkgs;
 
 mkShell {
@@ -28,10 +31,9 @@ mkShell {
 
   buildInputs = [
     haskell.compiler."ghc${ghcVersion}"
-    haskell-language-server
     cabal-install
     stack-wrapped
     nix
     cabal-docspec
-  ];
+  ] ++ brokenIn92;
 }

--- a/src/Control/Functor/Linear/Internal/Kan.hs
+++ b/src/Control/Functor/Linear/Internal/Kan.hs
@@ -98,13 +98,13 @@ instance Functor (Yoneda f) where
 instance Applicative f => Data.Applicative (Yoneda f) where
   pure a = Yoneda (\f -> pure (f a))
   {-# INLINE pure #-}
-  Yoneda m <*> Yoneda n = Yoneda (\f -> m (f .) <*> n id)
+  Yoneda m <*> Yoneda n = Yoneda (\f -> m (\g -> f . g) <*> n id)
   {-# INLINE (<*>) #-}
 
 instance Applicative f => Applicative (Yoneda f) where
   pure a = Yoneda (\f -> pure (f a))
   {-# INLINE pure #-}
-  Yoneda m <*> Yoneda n = Yoneda (\f -> m (f .) <*> n id)
+  Yoneda m <*> Yoneda n = Yoneda (\f -> m (\g -> f . g) <*> n id)
   {-# INLINE (<*>) #-}
 
 lowerYoneda :: Yoneda f a %1 -> f a
@@ -117,5 +117,5 @@ liftCurriedYonedaC fa = Curried (`yap` fa)
 {-# INLINE liftCurriedYonedaC #-}
 
 yap :: Applicative f => Yoneda f (a %1 -> b) %1 -> f a %1 -> Yoneda f b
-yap (Yoneda k) fa = Yoneda (\ab_r -> k (ab_r .) <*> fa)
+yap (Yoneda k) fa = Yoneda (\ab_r -> k (\g -> ab_r . g) <*> fa)
 {-# INLINE yap #-}

--- a/src/Data/Arity/Linear/Internal.hs
+++ b/src/Data/Arity/Linear/Internal.hs
@@ -1,13 +1,56 @@
 {-# LANGUAGE DataKinds #-}
+{-# LANGUAGE FlexibleInstances #-}
 {-# LANGUAGE LinearTypes #-}
+{-# LANGUAGE MultiParamTypeClasses #-}
+{-# LANGUAGE StandaloneKindSignatures #-}
 {-# LANGUAGE TypeFamilies #-}
 {-# LANGUAGE TypeOperators #-}
 {-# LANGUAGE UndecidableInstances #-}
 
-module Data.Arity.Linear.Internal (Arity) where
+module Data.Arity.Linear.Internal where
 
+import Data.Kind
 import GHC.TypeLits
+import GHC.Types
 
+data Peano = Z | S Peano
+
+type NatToPeano :: Nat -> Peano
+type family NatToPeano n where
+  NatToPeano 0 = 'Z
+  NatToPeano n = 'S (NatToPeano (n - 1))
+
+type PeanoToNat :: Peano -> Nat
+type family PeanoToNat n where
+  PeanoToNat 'Z = 0
+  PeanoToNat ('S n) = 1 + PeanoToNat n
+
+type FunN :: Peano -> Type -> Type -> Type
+type family FunN n a b where
+  FunN 'Z _ b = b
+  FunN ('S n) a b = a %1 -> FunN n a b
+
+-- | The 'Arity' type family exists to help the type checker fill in
+-- blanks. Chances are that you can safely ignore 'Arity' completely if it's in
+-- the type of a function you care. But read on if you are curious.
+--
+-- The idea is that in a function like 'Data.Replicator.Linear.elim' some of the
+-- type arguments are redundant. The function has an ambiguous type, so you will
+-- always have to help the compiler either with a type annotation or a type
+-- application. But there are several complete ways to do so. In
+-- 'Data.Replicator.Linear.elim', if you give the values of `n`, `a`, and `b`,
+-- then you can deduce the value of `f` (indeed, it's @'FunN' n a b@). With
+-- 'Arity' we can go in the other direction: if `b` and `f` are both known, then
+-- we know that `n` is @'Arity' b f@
+--
+-- 'Arity' returns a 'Nat' rather than a 'Peano' because the result is never
+-- consumed. It exists to infer arguments to functions such as
+-- 'Data.Replicator.Linear.elim' from the other arguments if they are known.
+--
+-- 'Arity' could be an associated type family to the 'IsFunN' type class. But
+-- it's better to make it a closed type family (which can't be associated to a
+-- type class) because it lets us give a well-defined error case.
+type Arity :: Type -> Type -> Nat
 type family Arity b f where
   Arity b b = 0
   Arity b (a %1 -> f) = Arity b f + 1
@@ -19,3 +62,26 @@ type family Arity b f where
           ':<>: 'ShowType b
           ':<>: 'Text "."
       )
+
+-- | The 'IsFun' type class is meant to help the type checker fill in
+-- blanks. Chances are that you can safely ignore 'IsFun' completely if it's in
+-- the type of a function you care. But read on if you are curious.
+--
+-- The type class 'IsFun' is a kind of inverse to 'FunN', it is meant to be
+-- read as @'IsFunN' a b f@ if and only if there exists `n` such that @f =
+-- 'FunN' n a b@ (`n` can be retrieved as @'Arity' b f@).
+--
+-- The reason why 'Arity' (read its documentation first) is not sufficient for
+-- our purpose, is that it can find `n` /if/ `f` is a linear function of the
+-- appropriate shape. But what if `f` is partially undetermined? Then it is
+-- likely that 'Arity' will be stuck. But we know, for instance, that if `f = a1
+-- %1 -> a2 %1 -> c` then we must have `a1 ~ a2`. The trick is that instance
+-- resolution of 'IsFun' will add unification constraints that the type checker
+-- has to solve. Look in particular at the instance @'IsFunN' a b (a' %p ->
+-- f))@: it matches liberally, so triggers on quite underdetermined `f`, but has
+-- equality constraints in its context which will help the type checker.
+class IsFunN a b f
+
+instance IsFunN a b b
+
+instance (IsFunN a b f, a' ~ a, p ~ 'One) => IsFunN a b (a' %p -> f)

--- a/src/Data/Replicator/Linear/Internal.hs
+++ b/src/Data/Replicator/Linear/Internal.hs
@@ -1,13 +1,15 @@
+{-# LANGUAGE AllowAmbiguousTypes #-}
 {-# LANGUAGE DataKinds #-}
 {-# LANGUAGE FlexibleContexts #-}
 {-# LANGUAGE FlexibleInstances #-}
-{-# LANGUAGE FunctionalDependencies #-}
 {-# LANGUAGE GADTs #-}
 {-# LANGUAGE LambdaCase #-}
 {-# LANGUAGE LinearTypes #-}
 {-# LANGUAGE MagicHash #-}
+{-# LANGUAGE MultiParamTypeClasses #-}
 {-# LANGUAGE ScopedTypeVariables #-}
 {-# LANGUAGE StandaloneKindSignatures #-}
+{-# LANGUAGE TypeApplications #-}
 {-# LANGUAGE TypeFamilies #-}
 {-# LANGUAGE TypeOperators #-}
 {-# LANGUAGE UnboxedTuples #-}
@@ -116,47 +118,47 @@ extract (Streamed (ReplicationStream s give _ _)) = give s
 extend :: (Replicator a %1 -> b) -> Replicator a %1 -> Replicator b
 extend f = map f . duplicate
 
+-- | Takes a function of type @a %1 -> a %1 -> ... %1 -> a %1 -> b@, and
+-- returns a @b@ . The replicator is used to supply all the items of type @a@
+-- required by the function.
+--
+-- For instance:
+--
+-- > elim @1 :: (a %1 -> b) %1 -> Replicator a %1 -> b
+-- > elim @2 :: (a %1 -> a %1 -> b) %1 -> Replicator a %1 -> b
+-- > elim @3 :: (a %1 -> a %1 -> a %1 -> b) %1 -> Replicator a %1 -> b
+--
+-- It is not always necessary to give the arity argument. It can be
+-- inferred from the function argument.
+--
+-- > elim (,) :: Replicator a %1 -> (a, a)
+-- > elim (,,) :: Replicator a %1 -> (a, a, a)
+elim :: forall (n :: Nat) a b f. (Elim (NatToPeano n) a b, IsFunN a b f, f ~ FunN (NatToPeano n) a b, n ~ Arity b f) => f %1 -> Replicator a %1 -> b
+elim f r = elim' @(NatToPeano n) f r
+
 -- | @'Elim' n a b f@ asserts that @f@ is a function taking @n@ linear arguments
 -- of type @a@ and then returning a value of type @b@.
 --
 -- It is solely used to define the type of the 'elim' function.
-type Elim :: Nat -> Type -> Type -> Type -> Constraint
-class (n ~ Arity b f) => Elim n a b f | n a b -> f, f b -> n where
-  -- | Takes a function of type @a %1 -> a %1 -> ... %1 -> a %1 -> b@, and
-  -- returns a @b@ . The replicator is used to supply all the items of type @a@
-  -- required by the function.
-  --
-  -- For instance:
-  --
-  -- > elim @1 :: (a %1 -> b) %1 -> Replicator a %1 -> b
-  -- > elim @2 :: (a %1 -> a %1 -> b) %1 -> Replicator a %1 -> b
-  -- > elim @3 :: (a %1 -> a %1 -> a %1 -> b) %1 -> Replicator a %1 -> b
-  --
-  -- It is not always necessary to give the arity argument. It can be
-  -- inferred from the function argument.
-  --
-  -- > elim (,) :: Replicator a %1 -> (a, a)
-  -- > elim (,,) :: Replicator a %1 -> (a, a, a)
-  elim :: f %1 -> Replicator a %1 -> b
-  elim f r = elim' f r
+type Elim :: Peano -> Type -> Type -> Constraint
+class Elim n a b where
+  -- Note that 'elim' is, in particular, used to force eta-expansion of
+  -- 'elim\''.  Otherwise, 'elim\'' might not get inlined (see
+  -- https://github.com/tweag/linear-base/issues/369).
+  elim' :: FunN n a b %1 -> Replicator a %1 -> b
 
-  -- 'elim' is used to force eta-expansion of 'elim\''.
-  -- Otherwise, 'elim\'' might not get inlined
-  -- (see https://github.com/tweag/linear-base/issues/369).
-  elim' :: f %1 -> Replicator a %1 -> b
-
-instance Elim 0 a b b where
+instance Elim 'Z a b where
   elim' b r =
     consume r & \case
       () -> b
   {-# INLINE elim' #-}
 
-instance (Arity b (a %1 -> b) ~ 1) => Elim 1 a b (a %1 -> b) where
+instance Elim ('S 'Z) a b where
   elim' f r = f (extract r)
   {-# INLINE elim' #-}
 
-instance {-# OVERLAPPABLE #-} (n ~ Arity b (a %1 -> f), Elim (n - 1) a b f) => Elim n a b (a %1 -> f) where
+instance (Elim ('S n) a b) => Elim ('S ('S n)) a b where
   elim' g r =
     next r & \case
-      (a, r') -> elim' (g a) r'
+      (a, r') -> elim' @('S n) (g a) r'
   {-# INLINE elim' #-}

--- a/src/Data/Unrestricted/Linear/Internal/Dupable.hs
+++ b/src/Data/Unrestricted/Linear/Internal/Dupable.hs
@@ -3,6 +3,7 @@
 {-# LANGUAGE LinearTypes #-}
 {-# LANGUAGE PolyKinds #-}
 {-# LANGUAGE ScopedTypeVariables #-}
+{-# LANGUAGE TypeApplications #-}
 {-# LANGUAGE NoImplicitPrelude #-}
 {-# OPTIONS_HADDOCK hide #-}
 

--- a/test/Test/Data/V.hs
+++ b/test/Test/Data/V.hs
@@ -17,7 +17,7 @@ vInspectionTests :: TestTree
 vInspectionTests =
   testGroup
     "Inspection testing of elim and make for V"
-    [ $(inspectTest $ 'make3 === 'manualMake3),
+    [ $(inspectTest $ 'make3 ==- 'manualMake3),
       $(inspectTest $ 'elim3 ==- 'manualElim3)
     ]
 

